File: experimental/java/src/main/java/com/azure/data/cosmos/serialization/hybridrow/layouts/LayoutCompiler.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.azure.data.cosmos.serialization.hybridrow.layouts;

import com.azure.data.cosmos.core.Out;
import com.azure.data.cosmos.serialization.hybridrow.SchemaId;
import com.azure.data.cosmos.serialization.hybridrow.schemas.ArrayPropertyType;
import com.azure.data.cosmos.serialization.hybridrow.schemas.MapPropertyType;
import com.azure.data.cosmos.serialization.hybridrow.schemas.Namespace;
import com.azure.data.cosmos.serialization.hybridrow.schemas.ObjectPropertyType;
import com.azure.data.cosmos.serialization.hybridrow.schemas.PrimitivePropertyType;
import com.azure.data.cosmos.serialization.hybridrow.schemas.Property;
import com.azure.data.cosmos.serialization.hybridrow.schemas.PropertyType;
import com.azure.data.cosmos.serialization.hybridrow.schemas.Schema;
import com.azure.data.cosmos.serialization.hybridrow.schemas.ScopePropertyType;
import com.azure.data.cosmos.serialization.hybridrow.schemas.SetPropertyType;
import com.azure.data.cosmos.serialization.hybridrow.schemas.TaggedPropertyType;
import com.azure.data.cosmos.serialization.hybridrow.schemas.TuplePropertyType;
import com.azure.data.cosmos.serialization.hybridrow.schemas.TypeKind;
import com.azure.data.cosmos.serialization.hybridrow.schemas.UdtPropertyType;
import com.google.common.base.Strings;
import org.checkerframework.checker.index.qual.NonNegative;

import javax.annotation.Nonnull;
import java.util.List;
import java.util.Optional;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.common.base.Strings.lenientFormat;

/**
 * Converts a logical schema into a physical layout.
 */
public final class LayoutCompiler {
    /**
     * Compiles a logical schema into a physical layout that can be used to read and write rows
     *
     * @param namespace     The namespace within which {@code schema} is defined
     * @param schema The logical schema to produce a layout for
     * @return The layout for the schema
     */
    @Nonnull
    public static Layout compile(@Nonnull final Namespace namespace, @Nonnull final Schema schema) {

        checkNotNull(namespace, "expected non-null namespace");
        checkNotNull(schema, "expected non-null schema");
        checkArgument(schema.type() == TypeKind.SCHEMA);
        checkArgument(!Strings.isNullOrEmpty(schema.name()));
        checkArgument(namespace.schemas().contains(schema));

        LayoutBuilder builder = new LayoutBuilder(schema.name(), schema.schemaId());
        LayoutCompiler.addProperties(builder, namespace, LayoutCode.SCHEMA, schema.properties());

        return builder.build();
    }

    private static void addProperties(
        @Nonnull final LayoutBuilder builder,
        @Nonnull final Namespace namespace,
        @Nonnull LayoutCode layoutCode,
        @Nonnull List<Property> properties) {

        final Out<TypeArgumentList> typeArgs = new Out<>();

        for (Property p : properties) {

            LayoutType type = LayoutCompiler.logicalToPhysicalType(namespace, p.type(), typeArgs);

            switch (LayoutCodeTraits.clearImmutableBit(type.layoutCode())) {

                case OBJECT_SCOPE: {
                    if (!p.type().nullable()) {
                        throw new LayoutCompilationException("Non-nullable sparse column are not supported.");
                    }
                    ObjectPropertyType op = (ObjectPropertyType)p.type();
                    builder.addObjectScope(p.path(), type);
                    LayoutCompiler.addProperties(builder, namespace, type.layoutCode(), op.properties());
                    builder.EndObjectScope();
                    break;
                }

                case ARRAY_SCOPE:
                case TYPED_ARRAY_SCOPE:
                case SET_SCOPE:
                case TYPED_SET_SCOPE:
                case MAP_SCOPE:
                case TYPED_MAP_SCOPE:
                case TUPLE_SCOPE:
                case TYPED_TUPLE_SCOPE:
                case TAGGED_SCOPE:
                case TAGGED2_SCOPE:
                case SCHEMA: {
                    if (!p.type().nullable()) {
                        throw new LayoutCompilationException("Non-nullable sparse column are not supported.");
                    }
                    builder.addTypedScope(p.path(), type, typeArgs.get());
                    break;
                }

                case NULLABLE_SCOPE: {
                    throw new LayoutCompilationException("Nullables cannot be explicitly declared as columns.");
                }

                default: {

                    if (p.type() instanceof PrimitivePropertyType) {

                        PrimitivePropertyType pp = (PrimitivePropertyType) p.type();

                        switch (pp.storage()) {

                            case FIXED:
                                if (LayoutCodeTraits.clearImmutableBit(layoutCode) != LayoutCode.SCHEMA) {
                                    throw new LayoutCompilationException(
                                        "Cannot have fixed storage within a sparse layoutCode.");
                                }
                                if (type.isNull() && !pp.nullable()) {
                                    throw new LayoutCompilationException(
                                        "Non-nullable null columns are not supported.");
                                }
                                builder.addFixedColumn(p.path(), type, pp.nullable(), pp.length());
                                break;

                            case VARIABLE:
                                if (LayoutCodeTraits.clearImmutableBit(layoutCode) != LayoutCode.SCHEMA) {
                                    throw new LayoutCompilationException(
                                        "Cannot have variable storage within a sparse layoutCode.");
                                }
                                if (!pp.nullable()) {
                                    throw new LayoutCompilationException(
                                        "Non-nullable variable columns are not supported.");
                                }
                                builder.addVariableColumn(p.path(), type, pp.length());
                                break;

                            case SPARSE:
                                if (!pp.nullable()) {
                                    throw new LayoutCompilationException(
                                        "Non-nullable sparse columns are not supported.");
                                }
                                builder.addSparseColumn(p.path(), type);
                                break;

                            default:
                                throw new LayoutCompilationException(
                                    lenientFormat("Unknown storage specification: %s", pp.storage()));
                        }
                    } else {
                        throw new LayoutCompilationException(
                            lenientFormat("Unknown property type: %s", type.name()));
                    }

                    break;
                }
            }
        }
    }

    private static LayoutType logicalToPhysicalType(
        Namespace namespace, PropertyType logicalType, Out<TypeArgumentList> typeArgs) {

        typeArgs.set(TypeArgumentList.EMPTY);
        boolean immutable = logicalType instanceof ScopePropertyType && ((ScopePropertyType) logicalType).immutable();

        switch (logicalType.type()) {

            case NULL:
                return LayoutTypes.NULL;

            case BOOLEAN:
                return LayoutTypes.BOOLEAN;

            case INT_8:
                return LayoutTypes.INT_8;

            case INT_16:
                return LayoutTypes.INT_16;

            case INT_32:
                return LayoutTypes.INT_32;

            case INT_64:
                return LayoutTypes.INT_64;

            case UINT_8:
                return LayoutTypes.UINT_8;

            case UINT_16:
                return LayoutTypes.UINT_16;

            case UINT_32:
                return LayoutTypes.UINT_32;

            case UINT_64:
                return LayoutTypes.UINT_64;

            case FLOAT_32:
                return LayoutTypes.FLOAT_32;

            case FLOAT_64:
                return LayoutTypes.FLOAT_64;

            case FLOAT_128:
                return LayoutTypes.FLOAT_128;

            case DECIMAL:
                return LayoutTypes.DECIMAL;

            case DATE_TIME:
                return LayoutTypes.DATE_TIME;

            case UNIX_DATE_TIME:
                return LayoutTypes.UNIX_DATE_TIME;

            case GUID:
                return LayoutTypes.GUID;

            case MONGODB_OBJECT_ID:
                throw new UnsupportedOperationException();
                // return LayoutTypes.MONGO_DB_OBJECT_ID;

            case UTF_8:
                return LayoutTypes.UTF_8;

            case BINARY:
                return LayoutTypes.BINARY;

            case VAR_INT:
                return LayoutTypes.VAR_INT;

            case VAR_UINT:
                return LayoutTypes.VAR_UINT;

            case OBJECT:
                return immutable ? LayoutTypes.IMMUTABLE_OBJECT : LayoutTypes.OBJECT;

            case ARRAY: {

                assert logicalType instanceof ArrayPropertyType;
                ArrayPropertyType ap = (ArrayPropertyType) logicalType;

                if (ap.items() != null && (ap.items().type() != TypeKind.ANY)) {

                    final Out<TypeArgumentList> out = new Out<>();

                    LayoutType itemType = LayoutCompiler.logicalToPhysicalType(namespace, ap.items(), out);
                    TypeArgumentList itemTypeArgs = out.get();

                    if (ap.items().nullable()) {
                        itemTypeArgs = new TypeArgumentList(new TypeArgument(itemType, itemTypeArgs));
                        itemType = itemType.isImmutable() ? LayoutTypes.IMMUTABLE_NULLABLE : LayoutTypes.NULLABLE;
                    }

                    typeArgs.set(new TypeArgumentList(new TypeArgument(itemType, itemTypeArgs)));
                    return immutable ? LayoutTypes.IMMUTABLE_TYPED_ARRAY : LayoutTypes.TYPED_ARRAY;
                }

                return immutable ? LayoutTypes.IMMUTABLE_ARRAY : LayoutTypes.ARRAY;
            }
            case SET: {

                assert logicalType instanceof SetPropertyType;
                SetPropertyType sp = (SetPropertyType) logicalType;

                if ((sp.items() != null) && (sp.items().type() != TypeKind.ANY)) {

                    final Out<TypeArgumentList> out = new Out<>();

                    LayoutType itemType = LayoutCompiler.logicalToPhysicalType(namespace, sp.items(), out);
                    TypeArgumentList itemTypeArgs = out.get();

                    if (sp.items().nullable()) {
                        itemTypeArgs = new TypeArgumentList(new TypeArgument(itemType, itemTypeArgs));
                        itemType = itemType.isImmutable() ? LayoutTypes.IMMUTABLE_NULLABLE : LayoutTypes.NULLABLE;
                    }

                    typeArgs.set(new TypeArgumentList(new TypeArgument(itemType, itemTypeArgs)));
                    return immutable ? LayoutTypes.IMMUTABLE_TYPED_SET : LayoutTypes.TYPED_SET;
                }

                // TODO(283638): implement sparse set

                throw new LayoutCompilationException(lenientFormat(
                    "Unknown property type: %s",
                    logicalType.type()
                ));
            }
            case MAP: {

                assert logicalType instanceof MapPropertyType;
                MapPropertyType mp = (MapPropertyType) logicalType;

                if (mp.keys() != null && (mp.keys().type() != TypeKind.ANY) && (mp.values() != null) && (mp.values().type() != TypeKind.ANY)) {

                    final Out<TypeArgumentList> out = new Out<>();

                    LayoutType keyType = LayoutCompiler.logicalToPhysicalType(namespace, mp.keys(), out);
                    TypeArgumentList keyTypeArgs = out.get();

                    if (mp.keys().nullable()) {
                        keyTypeArgs = new TypeArgumentList(new TypeArgument(keyType, keyTypeArgs));
                        keyType = keyType.isImmutable() ? LayoutTypes.IMMUTABLE_NULLABLE : LayoutTypes.NULLABLE;
                    }

                    LayoutType valueType = LayoutCompiler.logicalToPhysicalType(namespace, mp.values(), out);
                    TypeArgumentList valueTypeArgs = out.get();

                    if (mp.values().nullable()) {
                        valueTypeArgs = new TypeArgumentList(new TypeArgument(valueType, valueTypeArgs));
                        valueType = valueType.isImmutable() ? LayoutTypes.IMMUTABLE_NULLABLE : LayoutTypes.NULLABLE;
                    }

                    typeArgs.set(new TypeArgumentList(
                        new TypeArgument(keyType, keyTypeArgs),
                        new TypeArgument(valueType, valueTypeArgs)
                    ));

                    return immutable ? LayoutTypes.IMMUTABLE_TYPED_MAP : LayoutTypes.TYPED_MAP;
                }

                // TODO(283638): implement sparse map

                throw new LayoutCompilationException(lenientFormat(
                    "Unknown property type: %s", logicalType.type())
                );
            }
            case TUPLE: {

                assert logicalType instanceof TuplePropertyType;
                final TuplePropertyType tp = (TuplePropertyType) logicalType;

                final TypeArgument[] args = new TypeArgument[tp.items().size()];
                final Out<TypeArgumentList> out = new Out<>();

                for (int i = 0; i < tp.items().size(); i++) {

                    LayoutType itemType = LayoutCompiler.logicalToPhysicalType(namespace, tp.items().get(i), out);
                    TypeArgumentList itemTypeArgs = out.get();

                    if (tp.items().get(i).nullable()) {
                        itemTypeArgs = new TypeArgumentList(new TypeArgument(itemType, itemTypeArgs));
                        itemType = itemType.isImmutable() ? LayoutTypes.IMMUTABLE_NULLABLE : LayoutTypes.NULLABLE;
                    }

                    args[i] = new TypeArgument(itemType, itemTypeArgs);
                }

                typeArgs.setAndGet(new TypeArgumentList(args));
                return immutable ? LayoutTypes.IMMUTABLE_TYPED_TUPLE : LayoutTypes.TYPED_TUPLE;
            }
            case TAGGED: {

                assert logicalType instanceof TaggedPropertyType;
                TaggedPropertyType tg = (TaggedPropertyType) logicalType;

                if (tg.items().size() < TaggedPropertyType.MIN_TAGGED_ARGUMENTS || (tg.items().size() > TaggedPropertyType.MAX_TAGGED_ARGUMENTS)) {
                    throw new LayoutCompilationException(lenientFormat(
                        "Invalid number of arguments in Tagged: %s <= %s <= %s",
                        TaggedPropertyType.MIN_TAGGED_ARGUMENTS,
                        tg.items().size(),
                        TaggedPropertyType.MAX_TAGGED_ARGUMENTS
                    ));
                }

                final Out<TypeArgumentList> out = new Out<>();
                final TypeArgument[] tgArgs = new TypeArgument[tg.items().size() + 1];

                tgArgs[0] = new TypeArgument(LayoutTypes.UINT_8, TypeArgumentList.EMPTY);

                for (int i = 0; i < tg.items().size(); i++) {

                    LayoutType itemType = LayoutCompiler.logicalToPhysicalType(namespace, tg.items().get(i), out);
                    TypeArgumentList itemTypeArgs = out.get();

                    if (tg.items().get(i).nullable()) {
                        itemTypeArgs = new TypeArgumentList(new TypeArgument(itemType, itemTypeArgs));
                        itemType = itemType.isImmutable() ? LayoutTypes.IMMUTABLE_NULLABLE : LayoutTypes.NULLABLE;
                    }

                    tgArgs[i + 1] = new TypeArgument(itemType, itemTypeArgs);
                }

                typeArgs.set(new TypeArgumentList(tgArgs));

                switch (tg.items().size()) {
                    case 1:
                        return immutable ? LayoutTypes.IMMUTABLE_TAGGED : LayoutTypes.TAGGED;
                    case 2:
                        return immutable ? LayoutTypes.IMMUTABLE_TAGGED_2 : LayoutTypes.TAGGED_2;
                    default:
                        throw new LayoutCompilationException("Unexpected tagged arity");
                }
            }
            case SCHEMA: {

                assert logicalType instanceof UdtPropertyType;
                UdtPropertyType up = (UdtPropertyType) logicalType;

                final Optional<Schema> udtSchema;

                if (up.schemaId() == SchemaId.NONE) {
                    udtSchema = namespace.schemas().stream()
                        .filter(schema -> up.name().equals(schema.name()))
                        .findFirst();
                } else {
                    udtSchema = namespace.schemas().stream()
                        .filter(schema -> up.schemaId().equals(schema.schemaId()))
                        .findFirst();
                    if (udtSchema.isPresent() && !up.name().equals(udtSchema.get().name())) {
                        throw new LayoutCompilationException(lenientFormat(
                            "ambiguous schema reference: '%s:%s'", up.name(), up.schemaId()
                        ));
                    }
                }

                if (!udtSchema.isPresent()) {
                    throw new LayoutCompilationException(lenientFormat(
                        "cannot resolve schema reference '%s:%s'", up.name(), up.schemaId()
                    ));
                }

                typeArgs.set(new TypeArgumentList(udtSchema.get().schemaId()));
                return immutable ? LayoutTypes.IMMUTABLE_UDT : LayoutTypes.UDT;
            }
            default:
                throw new LayoutCompilationException(Strings.lenientFormat(
                    "Unknown property type: %s", logicalType.type()
                ));
        }
    }
}

File: src/Serialization/HybridRow.Native/LayoutCompiler.cpp
// ------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
// ------------------------------------------------------------

#include "pch.h"
#include "Schema.h"
#include "Namespace.h"
#include "LayoutCode.h"
#include "LayoutType.h"
#include "LayoutBuilder.h"
#include "LayoutCompiler.h"
#include "ArrayPropertyType.h"
#include "LayoutCodeTraits.h"
#include "MapPropertyType.h"
#include "ObjectPropertyType.h"
#include "PrimitivePropertyType.h"
#include "SetPropertyType.h"
#include "TaggedPropertyType.h"
#include "TuplePropertyType.h"
#include "UdtPropertyType.h"

// ReSharper disable CppClangTidyCppcoreguidelinesProTypeStaticCastDowncast
namespace cdb_hr
{
  class LayoutBuilder;

  // Forward declaration.
  static void AddBase(LayoutBuilder& builder, const Namespace& ns, const Schema& s);
  static void AddProperties(LayoutBuilder& builder, SchemaLanguageVersion v, const Namespace& ns, LayoutCode scope,
                            const std::vector<std::unique_ptr<Property>>& properties);
  static std::tuple<const LayoutType*, TypeArgumentList> LogicalToPhysicalType(
    SchemaLanguageVersion v, const Namespace& ns, const PropertyType& logicalType);
  static const LayoutType* PrimitiveToPhysicalType(TypeKind type);
  constexpr static std::string_view BasePropertyName = "__base"sv;

  std::unique_ptr<Layout> LayoutCompiler::Compile(const Namespace& ns, const Schema& schema) noexcept(false)
  {
    cdb_core::Contract::Requires(schema.GetType() == TypeKind::Schema);
    cdb_core::Contract::Requires(!schema.GetName().empty());
    cdb_core::Contract::Requires(std::find_if(ns.GetSchemas().begin(), ns.GetSchemas().end(),
      [&schema](const std::unique_ptr<Schema>& p) { return p.get() == &schema; }) != ns.GetSchemas().end());

    SchemaLanguageVersion v = schema.GetEffectiveSdlVersion(ns);
    LayoutBuilder builder{schema.GetName(), schema.GetSchemaId()};
    if (!schema.GetBaseName().empty())
    {
      AddBase(builder, ns, schema);
    }
    AddProperties(builder, v, ns, LayoutCode::Schema, schema.GetProperties());

    return builder.Build();
  }

  void AddBase(LayoutBuilder& builder, const Namespace& ns, const Schema& s)
  {
    const auto& schemas = ns.GetSchemas();
    decltype(schemas.begin()) iter;
    if (s.GetBaseSchemaId() == SchemaId::Invalid())
    {
      iter = std::find_if(schemas.begin(), schemas.end(),
        [&s](const std::unique_ptr<Schema>& q) { return q->GetName() == s.GetBaseName(); });
    }
    else
    {
      iter = std::find_if(schemas.begin(), schemas.end(),
        [&s](const std::unique_ptr<Schema>& q) { return q->GetSchemaId() == s.GetBaseSchemaId(); });
    }

    if (iter == schemas.end())
    {
      throw LayoutCompiler::LayoutCompilationException(cdb_core::make_string("Cannot resolve schema reference: '%s:%d'",
        s.GetBaseName().data(), s.GetBaseSchemaId().Id()));
    }

    const Schema& bs = **iter;
    if (bs.GetName() != s.GetBaseName())
    {
      throw LayoutCompiler::LayoutCompilationException(cdb_core::make_string("Ambiguous schema reference: '%s:%d'",
        s.GetBaseName().data(), s.GetBaseSchemaId().Id()));
    }

    builder.AddTypedScope(BasePropertyName, &LayoutLiteral::UDT, bs.GetSchemaId());
  }

  void AddProperties(
    LayoutBuilder& builder,
    SchemaLanguageVersion v,
    const Namespace& ns,
    LayoutCode scope,
    const std::vector<std::unique_ptr<Property>>& properties)
  {
    for (const auto& ptr : properties)
    {
      const Property& p = *ptr;
      if (!p.GetPropertyType().has_value())
      {
        throw LayoutCompiler::LayoutCompilationException(cdb_core::make_string("Property missing type"));
      }
      const PropertyType& pt = p.GetPropertyType().value();
      auto [type, typeArgs] = LogicalToPhysicalType(v, ns, pt);
      switch (LayoutCodeTraits::ClearImmutableBit(type->GetLayoutCode()))
      {
      case LayoutCode::ObjectScope:
      {
        if (!pt.GetNullable())
        {
          throw LayoutCompiler::LayoutCompilationException("Non-nullable sparse column are not supported.");
        }

        cdb_core::Contract::Invariant(pt.GetKind() == PropertyKind::Object);
        const ObjectPropertyType& op = static_cast<const ObjectPropertyType&>(pt);
        builder.AddObjectScope(p.GetPath(), type);
        AddProperties(builder, v, ns, type->GetLayoutCode(), op.GetProperties());
        builder.EndObjectScope();
        break;
      }

      case LayoutCode::ArrayScope:
      case LayoutCode::TypedArrayScope:
      case LayoutCode::SetScope:
      case LayoutCode::TypedSetScope:
      case LayoutCode::MapScope:
      case LayoutCode::TypedMapScope:
      case LayoutCode::TupleScope:
      case LayoutCode::TypedTupleScope:
      case LayoutCode::TaggedScope:
      case LayoutCode::Tagged2Scope:
      case LayoutCode::Schema:
      {
        if (!pt.GetNullable())
        {
          throw LayoutCompiler::LayoutCompilationException("Non-nullable sparse column are not supported.");
        }

        builder.AddTypedScope(p.GetPath(), type, typeArgs);
        break;
      }

      case LayoutCode::NullableScope:
      {
        throw LayoutCompiler::LayoutCompilationException("Nullables cannot be explicitly declared as columns.");
      }

      default:
      {
        if (pt.GetKind() == PropertyKind::Primitive)
        {
          const PrimitivePropertyType& pp = static_cast<const PrimitivePropertyType&>(pt);
          if ((pp.GetType() == TypeKind::Enum) && (v < SchemaLanguageVersion::V2))
          {
            throw LayoutCompiler::LayoutCompilationException("Enums require SDL v2 or higher.");
          }

          switch (pp.GetStorage())
          {
          case StorageKind::Fixed:
            if (LayoutCodeTraits::ClearImmutableBit(scope) != LayoutCode::Schema)
            {
              throw LayoutCompiler::LayoutCompilationException("Cannot have fixed storage within a sparse scope.");
            }

            if (type->IsNull() && !pp.GetNullable())
            {
              throw LayoutCompiler::LayoutCompilationException("Non-nullable null columns are not supported.");
            }

            builder.AddFixedColumn(p.GetPath(), type, pp.GetNullable(), pp.GetLength());
            break;
          case StorageKind::Variable:
            if (pp.GetType() == TypeKind::Enum)
            {
              throw LayoutCompiler::LayoutCompilationException("Enums cannot have storage specification: Variable");
            }

            if (LayoutCodeTraits::ClearImmutableBit(scope) != LayoutCode::Schema)
            {
              throw LayoutCompiler::LayoutCompilationException("Cannot have variable storage within a sparse scope.");
            }

            if (!pp.GetNullable())
            {
              throw LayoutCompiler::LayoutCompilationException("Non-nullable variable columns are not supported.");
            }

            builder.AddVariableColumn(p.GetPath(), type, pp.GetLength());
            break;
          case StorageKind::Sparse:
            if (!pp.GetNullable())
            {
              throw LayoutCompiler::LayoutCompilationException("Non-nullable sparse columns are not supported.");
            }

            builder.AddSparseColumn(p.GetPath(), type);
            break;
          default:
            throw LayoutCompiler::LayoutCompilationException(cdb_core::make_string("Unknown storage specification: %u",
              pp.GetStorage()));
          }
        }
        else
        {
          throw LayoutCompiler::LayoutCompilationException(cdb_core::make_string("Unknown property type: %s",
            type->GetName().data()));
        }

        break;
      }
      }
    }
  }

  static std::tuple<const LayoutType*, TypeArgumentList> LogicalToPhysicalType(
    SchemaLanguageVersion v, const Namespace& ns, const PropertyType& logicalType)
  {
    bool immutable = (logicalType.GetKind() == PropertyKind::Primitive)
                       ? false
                       : static_cast<const ScopePropertyType&>(logicalType).GetImmutable();

    switch (logicalType.GetType())
    {
    case TypeKind::Null:
    case TypeKind::Boolean:
    case TypeKind::Int8:
    case TypeKind::Int16:
    case TypeKind::Int32:
    case TypeKind::Int64:
    case TypeKind::UInt8:
    case TypeKind::UInt16:
    case TypeKind::UInt32:
    case TypeKind::UInt64:
    case TypeKind::Float32:
    case TypeKind::Float64:
    case TypeKind::Float128:
    case TypeKind::Decimal:
    case TypeKind::DateTime:
    case TypeKind::UnixDateTime:
    case TypeKind::Guid:
    case TypeKind::MongoDbObjectId:
    case TypeKind::Utf8:
    case TypeKind::Binary:
    case TypeKind::VarInt:
    case TypeKind::VarUInt:
      return {PrimitiveToPhysicalType(logicalType.GetType()), {}};

    case TypeKind::Object:
      return {immutable ? &LayoutLiteral::ImmutableObject : &LayoutLiteral::Object, {}};
    case TypeKind::Array:
    {
      cdb_core::Contract::Invariant(logicalType.GetKind() == PropertyKind::Array);
      const ArrayPropertyType& ap = static_cast<const ArrayPropertyType&>(logicalType);
      if (ap.GetItems().has_value())
      {
        const PropertyType& pt = *ap.GetItems();
        if (pt.GetType() != TypeKind::Any)
        {
          auto [itemType, itemTypeArgs] = LogicalToPhysicalType(v, ns, pt);
          if (pt.GetNullable())
          {
            itemTypeArgs = {{itemType, itemTypeArgs}};
            itemType = itemType->IsImmutable() ? &LayoutLiteral::ImmutableNullable : &LayoutLiteral::Nullable;
          }

          TypeArgumentList typeArgs{{itemType, itemTypeArgs}};
          return {immutable ? &LayoutLiteral::ImmutableTypedArray : &LayoutLiteral::TypedArray, typeArgs};
        }
      }

      return {immutable ? &LayoutLiteral::ImmutableArray : &LayoutLiteral::Array, {}};
    }
    case TypeKind::Set:
    {
      cdb_core::Contract::Invariant(logicalType.GetKind() == PropertyKind::Set);
      const SetPropertyType& sp = static_cast<const SetPropertyType&>(logicalType);
      if (sp.GetItems().has_value())
      {
        const PropertyType& pt = *sp.GetItems();
        if (pt.GetType() != TypeKind::Any)
        {
          auto [itemType, itemTypeArgs] = LogicalToPhysicalType(v, ns, pt);
          if (pt.GetNullable())
          {
            itemTypeArgs = {{itemType, itemTypeArgs}};
            itemType = itemType->IsImmutable() ? &LayoutLiteral::ImmutableNullable : &LayoutLiteral::Nullable;
          }

          TypeArgumentList typeArgs{{itemType, itemTypeArgs}};
          return {immutable ? &LayoutLiteral::ImmutableTypedSet : &LayoutLiteral::TypedSet, typeArgs};
        }
      }

      // TODO(283638): implement sparse set.
      throw LayoutCompiler::LayoutCompilationException(cdb_core::make_string("Unknown property type: %u",
        logicalType.GetType()));
    }
    case TypeKind::Map:
    {
      cdb_core::Contract::Invariant(logicalType.GetKind() == PropertyKind::Map);
      const MapPropertyType& mp = static_cast<const MapPropertyType&>(logicalType);
      if (mp.GetKeys().has_value() && mp.GetValues().has_value())
      {
        const PropertyType& kpt = *mp.GetKeys();
        const PropertyType& vpt = *mp.GetValues();
        if ((kpt.GetType() != TypeKind::Any) && (vpt.GetType() != TypeKind::Any))
        {
          auto [keyType, keyTypeArgs] = LogicalToPhysicalType(v, ns, kpt);
          if (kpt.GetNullable())
          {
            keyTypeArgs = {{keyType, keyTypeArgs}};
            keyType = keyType->IsImmutable() ? &LayoutLiteral::ImmutableNullable : &LayoutLiteral::Nullable;
          }

          auto [valueType, valueTypeArgs] = LogicalToPhysicalType(v, ns, vpt);
          if (vpt.GetNullable())
          {
            valueTypeArgs = {{valueType, valueTypeArgs}};
            valueType = valueType->IsImmutable() ? &LayoutLiteral::ImmutableNullable : &LayoutLiteral::Nullable;
          }

          TypeArgumentList typeArgs{{{keyType, keyTypeArgs}, {valueType, valueTypeArgs}}};
          return {immutable ? &LayoutLiteral::ImmutableTypedMap : &LayoutLiteral::TypedMap, typeArgs};
        }
      }

      // TODO(283638): implement sparse map.
      throw LayoutCompiler::LayoutCompilationException(cdb_core::make_string("Unknown property type: %u",
        logicalType.GetType()));
    }
    case TypeKind::Tuple:
    {
      cdb_core::Contract::Invariant(logicalType.GetKind() == PropertyKind::Tuple);
      const TuplePropertyType& tp = static_cast<const TuplePropertyType&>(logicalType);
      tla::vector<TypeArgument> args{};
      args.reserve(tp.GetItems().size());
      for (const auto& item : tp.GetItems())
      {
        auto [itemType, itemTypeArgs] = LogicalToPhysicalType(v, ns, *item);
        if (item->GetNullable())
        {
          itemTypeArgs = {{itemType, itemTypeArgs}};
          itemType = itemType->IsImmutable() ? &LayoutLiteral::ImmutableNullable : &LayoutLiteral::Nullable;
        }

        args.emplace_back(itemType, itemTypeArgs);
      }

      TypeArgumentList typeArgs{args};
      return {immutable ? &LayoutLiteral::ImmutableTypedTuple : &LayoutLiteral::TypedTuple, typeArgs};
    }
    case TypeKind::Tagged:
    {
      cdb_core::Contract::Invariant(logicalType.GetKind() == PropertyKind::Tagged);
      const TaggedPropertyType& tg = static_cast<const TaggedPropertyType&>(logicalType);
      if ((tg.GetItems().size() < TaggedPropertyType::MinTaggedArguments) ||
        (tg.GetItems().size() > TaggedPropertyType::MaxTaggedArguments))
      {
        throw LayoutCompiler::LayoutCompilationException(
          cdb_core::make_string("Invalid number of arguments in Tagged: %u <= %u <= %u",
            TaggedPropertyType::MinTaggedArguments,
            tg.GetItems().size(),
            TaggedPropertyType::MaxTaggedArguments));
      }

      tla::vector<TypeArgument> tgArgs{};
      tgArgs.reserve(tg.GetItems().size() + 1);
      tgArgs.emplace_back(&LayoutLiteral::UInt8);
      for (const auto& item : tg.GetItems())
      {
        auto [itemType, itemTypeArgs] = LogicalToPhysicalType(v, ns, *item);
        if (item->GetNullable())
        {
          itemTypeArgs = {{itemType, itemTypeArgs}};
          itemType = itemType->IsImmutable() ? &LayoutLiteral::ImmutableNullable : &LayoutLiteral::Nullable;
        }

        tgArgs.emplace_back(itemType, itemTypeArgs);
      }

      TypeArgumentList typeArgs{tgArgs};
      switch (tg.GetItems().size())
      {
      case 1:
        return {immutable ? &LayoutLiteral::ImmutableTagged : &LayoutLiteral::Tagged, typeArgs};
      case 2:
        return {immutable ? &LayoutLiteral::ImmutableTagged2 : &LayoutLiteral::Tagged2, typeArgs};
      default:
        throw LayoutCompiler::LayoutCompilationException("Unexpected tagged arity");
      }
    }
    case TypeKind::Schema:
    {
      cdb_core::Contract::Invariant(logicalType.GetKind() == PropertyKind::Udt);
      const UdtPropertyType& up = static_cast<const UdtPropertyType&>(logicalType);
      const auto& schemas = ns.GetSchemas();
      decltype(schemas.begin()) iter;
      if (up.GetSchemaId() == SchemaId::Invalid())
      {
        iter = std::find_if(schemas.begin(), schemas.end(),
          [&up](const std::unique_ptr<Schema>& s) { return s->GetName() == up.GetName(); });
      }
      else
      {
        iter = std::find_if(schemas.begin(), schemas.end(),
          [&up](const std::unique_ptr<Schema>& s) { return s->GetSchemaId() == up.GetSchemaId(); });
      }
      if (iter == schemas.end())
      {
        throw LayoutCompiler::LayoutCompilationException(cdb_core::make_string(
          "Cannot resolve schema reference: '%s:%d'",
          up.GetName().data(), up.GetSchemaId().Id()));
      }

      const Schema& udtSchema = **iter;
      if (udtSchema.GetName() != up.GetName())
      {
        throw LayoutCompiler::LayoutCompilationException(cdb_core::make_string("Ambiguous schema reference: '%s:%d'",
          up.GetName().data(), up.GetSchemaId().Id()));
      }

      TypeArgumentList typeArgs{udtSchema.GetSchemaId()};
      return {immutable ? &LayoutLiteral::ImmutableUDT : &LayoutLiteral::UDT, typeArgs};
    }
    case TypeKind::Enum:
    {
      if (v < SchemaLanguageVersion::V2)
      {
        throw LayoutCompiler::LayoutCompilationException("Enums require SDL v2 or higher.");
      }

      cdb_core::Contract::Invariant(logicalType.GetKind() == PropertyKind::Primitive);
      const PrimitivePropertyType& ep = static_cast<const PrimitivePropertyType&>(logicalType);
      const auto& enums = ns.GetEnums();
      decltype(enums.begin()) iter = std::find_if(enums.begin(), enums.end(),
        [&ep](const std::unique_ptr<EnumSchema>& es) { return es->GetName() == ep.GetEnum(); });
      if (iter == enums.end())
      {
        throw LayoutCompiler::LayoutCompilationException(cdb_core::make_string(
          "Cannot resolve enum schema reference: '%s'",
          ep.GetEnum().data()));
      }

      const EnumSchema& enumSchema = **iter;
      return {PrimitiveToPhysicalType(enumSchema.GetType()), {}};
    }
    default:
      throw LayoutCompiler::LayoutCompilationException(cdb_core::make_string("Unknown property type: %u",
        logicalType.GetType()));
    }
  }

  static const LayoutType* PrimitiveToPhysicalType(TypeKind type)
  {
    switch (type)
    {
    case TypeKind::Null:
      return &LayoutLiteral::Null;
    case TypeKind::Boolean:
      return &LayoutLiteral::Boolean;
    case TypeKind::Int8:
      return &LayoutLiteral::Int8;
    case TypeKind::Int16:
      return &LayoutLiteral::Int16;
    case TypeKind::Int32:
      return &LayoutLiteral::Int32;
    case TypeKind::Int64:
      return &LayoutLiteral::Int64;
    case TypeKind::UInt8:
      return &LayoutLiteral::UInt8;
    case TypeKind::UInt16:
      return &LayoutLiteral::UInt16;
    case TypeKind::UInt32:
      return &LayoutLiteral::UInt32;
    case TypeKind::UInt64:
      return &LayoutLiteral::UInt64;
    case TypeKind::Float32:
      return &LayoutLiteral::Float32;
    case TypeKind::Float64:
      return &LayoutLiteral::Float64;
    case TypeKind::Float128:
      return &LayoutLiteral::Float128;
    case TypeKind::Decimal:
      return &LayoutLiteral::Decimal;
    case TypeKind::DateTime:
      return &LayoutLiteral::DateTime;
    case TypeKind::UnixDateTime:
      return &LayoutLiteral::UnixDateTime;
    case TypeKind::Guid:
      return &LayoutLiteral::Guid;
    case TypeKind::MongoDbObjectId:
      return &LayoutLiteral::MongoDbObjectId;
    case TypeKind::Utf8:
      return &LayoutLiteral::Utf8;
    case TypeKind::Binary:
      return &LayoutLiteral::Binary;
    case TypeKind::VarInt:
      return &LayoutLiteral::VarInt;
    case TypeKind::VarUInt:
      return &LayoutLiteral::VarUInt;
    default:
      throw LayoutCompiler::LayoutCompilationException(cdb_core::make_string("Unknown property type: %u", type));
    }
  }
}
